Mask R-CNN

cabelo opened the pull request:
<cut/>

![img01vis _mask_rcnn_out](https://user-images.githubusercontent.com/675645/52387238-91a93500-2a70-11e9-86e1-d5ad72af320d.jpg)

```
**WIP**
force_builders=linux,windows,mac,docs
test_modules=none
```

dkurt commented on samples/dnn/mask_rcnn.cpp:
```diff
@@ -0,0 +1,233 @@
+int main(int argc, char** argv)
+{
+    CommandLineParser parser(argc, argv, keys);
+    parser.about("Use this script to run object detection using YOLO3 in OpenCV.");
```
Please update a comment. This sample is not for YOLO.

cabelo replied:
Sorry, done...

dkurt commented on samples/dnn/mask_rcnn.cpp:
```diff
@@ -0,0 +1,233 @@
+    maskThreshold = parser.get<float>("mthr");
+
+    confThreshold = 0.5;
+    maskThreshold = 0.3;
```
In this way we cannot specify custom thresholds.

cabelo replied:
Done

dkurt commented on samples/dnn/mask_rcnn.cpp:
```diff
@@ -0,0 +1,233 @@
+        vector<double> layersTimes;
+        double freq = getTickFrequency() / 1000;
+        double t = net.getPerfProfile(layersTimes) / freq;
+        string label = format("openCV with Inference - cabelo@opensuse.org : %0.0f ms", t);
```
You can move a contact info to the comments block in the header.

cabelo replied:
Done

dkurt commented on samples/dnn/mask_rcnn.cpp:
```diff
@@ -0,0 +1,233 @@
+
+    try {
+
+        outputFile = "mask_rcnn_out_cpp.avi";
```
Let's keep a sample as minimal as it possible. Just preview output image but not write an output files.

cabelo replied:
Done

dkurt commented on samples/dnn/mask_rcnn.cpp:
```diff
@@ -0,0 +1,233 @@
+    }
+
+    String textGraph = "./mask_rcnn_inception_v2_coco_2018_01_28.pbtxt";
+    String modelWeights = "./mask_rcnn_inception_v2_coco_2018_01_28/frozen_inference_graph.pb";
```
Please move all the hardcoded arguments to command line arguments.

cabelo replied:
Done...

dkurt commented on samples/dnn/mask_rcnn.cpp:
```diff
@@ -0,0 +1,219 @@
+int main(int argc, char** argv)
+{
+    CommandLineParser parser(argc, argv, keys);
+    parser.about("Use this script to run object detection using Segmentation using Mask R-CNN in OpenCV.");
```
> parser.about("Use this script to run object detection using Segmentation using Mask R-CNN in OpenCV.");

It sounds a bit weird. This sample demonstrates **instance segmentation** network called Mask-RCNN.

cabelo replied:
Done!

dkurt commented on samples/dnn/mask_rcnn.cpp:
```diff
@@ -0,0 +1,219 @@
+    String textGraph = parser.get<string>("textgraph");
+    String modelWeights = parser.get<string>("weights");
+
+    string classesFile = "mscoco_labels.names";
```
Please align names of arguments with a [python version](https://github.com/opencv/opencv/blob/f67b197d496c602969b89dc90adc70418ae243e8/samples/dnn/mask_rcnn.py) of this sample.

cabelo replied:
Ok !

cabelo replied:
Done

Daniil-Osokin commented on samples/dnn/mask_rcnn.cpp:
```diff
@@ -0,0 +1,220 @@
+
+    confThreshold = parser.get<float>("cthr");
+    maskThreshold = parser.get<float>("mthr");
+    String textGraph = parser.get<string>("config");
```
Suggest to use just one string type - `string`.

cabelo replied:
done

Daniil-Osokin commented on samples/dnn/mask_rcnn.cpp:
```diff
@@ -0,0 +1,220 @@
+
+    ifstream ifs(classesFile.c_str());
+    string line;
+    while (getline(ifs, line)) classes.push_back(line);
```
Since this demo linked to Tensorflow OD API, suggest to pre-definne coco classes and use them. Now one have to create this file (in which format?) and guess classes order, so looks not very usable.

Daniil-Osokin commented on samples/dnn/mask_rcnn.cpp:
```diff
@@ -0,0 +1,220 @@
+    string line;
+    while (getline(ifs, line)) classes.push_back(line);
+
+    string colorsFile = "colors.txt";
```
To eliminate the user need to create this file, suggest to set the `seed` and initialize colors randomly.

Daniil-Osokin commented on samples/dnn/mask_rcnn.cpp:
```diff
@@ -0,0 +1,220 @@
+        colors.push_back(Scalar(r, g, b, 255.0));
+    }
+
+    Net net = readNetFromTensorflow(modelWeights, textGraph);
```
This can be simplified to `readNet`.

Daniil-Osokin commented on samples/dnn/mask_rcnn.cpp:
```diff
@@ -0,0 +1,220 @@
+
+    string str, outputFile;
+    VideoCapture cap;
+    VideoWriter video;
```
Extra object.

Daniil-Osokin commented on samples/dnn/mask_rcnn.cpp:
```diff
@@ -0,0 +1,220 @@
+    net.setPreferableBackend(DNN_BACKEND_OPENCV);
+    net.setPreferableTarget(DNN_TARGET_CPU);
+
+    string str, outputFile;
```
`outputFile` is extra.

Daniil-Osokin commented on samples/dnn/mask_rcnn.cpp:
```diff
@@ -0,0 +1,220 @@
+    Mat frame, blob;
+
+    try {
+
```
Extra line.

Daniil-Osokin commented on samples/dnn/mask_rcnn.cpp:
```diff
@@ -0,0 +1,220 @@
+            cap.open(str);
+        }
+        else cap.open(parser.get<int>("device"));
+
```
Extra line.

Daniil-Osokin commented on samples/dnn/mask_rcnn.cpp:
```diff
@@ -0,0 +1,220 @@
+        cout << "Could not open the input image/video stream" << endl;
+        return 0;
+    }
+
```
Extra lines.

Daniil-Osokin commented on samples/dnn/mask_rcnn.cpp:
```diff
@@ -0,0 +1,220 @@
+    }
+
+
+    static const string kWinName = "Deep learning object detection in OpenCV";
```
This not needed here. Just pass string to `imshow`.

Daniil-Osokin replied:
String should be clarified, there is already the object detection sample. This can be `Mask R-CNN sample`.

Daniil-Osokin commented on samples/dnn/mask_rcnn.cpp:
```diff
@@ -0,0 +1,220 @@
+    while (waitKey(1) < 0)
+    {
+        cap >> frame;
+
```
Extra line.

Daniil-Osokin commented on samples/dnn/mask_rcnn.cpp:
```diff
@@ -0,0 +1,220 @@
+        }
+
+        blobFromImage(frame, blob, 1.0, Size(frame.cols, frame.rows), Scalar(), true, false);
+
```
Extra line.

Daniil-Osokin commented on samples/dnn/mask_rcnn.cpp:
```diff
@@ -0,0 +1,220 @@
+        net.forward(outs, outNames);
+
+        postprocess(frame, outs);
+
```
Extra line.

Daniil-Osokin commented on samples/dnn/mask_rcnn.cpp:
```diff
@@ -0,0 +1,220 @@
+        vector<double> layersTimes;
+        double freq = getTickFrequency() / 1000;
+        double t = net.getPerfProfile(layersTimes) / freq;
+        string label = format("openCV Mask R-CNN : %0.0f ms", t);
```
`OpenCV`

Daniil-Osokin commented on samples/dnn/mask_rcnn.cpp:
```diff
@@ -0,0 +1,220 @@
+        double t = net.getPerfProfile(layersTimes) / freq;
+        string label = format("openCV Mask R-CNN : %0.0f ms", t);
+        putText(frame, label, Point(0, 15), FONT_HERSHEY_SIMPLEX, 0.5, Scalar(0, 0, 0));
+
```
Extra line.

Daniil-Osokin commented on samples/dnn/mask_rcnn.cpp:
```diff
@@ -0,0 +1,220 @@
+        Mat detectedFrame;
+        frame.convertTo(detectedFrame, CV_8U);
+        imshow(kWinName, frame);
+
```
Extra line.

Daniil-Osokin commented on samples/dnn/mask_rcnn.cpp:
```diff
@@ -0,0 +1,220 @@
+            top = max(0, min(top, frame.rows - 1));
+            right = max(0, min(right, frame.cols - 1));
+            bottom = max(0, min(bottom, frame.rows - 1));
+            Rect box = Rect(left, top, right - left + 1, bottom - top + 1);
```
Suggest
```
Rect box(Point(left, top), Point(right, bottom));
Rect imageBox(0, 0, frame.cols - 1, frame.rows - 1);
box &= imageBox;
```

Daniil-Osokin commented on samples/dnn/mask_rcnn.cpp:
```diff
@@ -0,0 +1,220 @@
+
+void drawBox(Mat& frame, int classId, float conf, Rect box, Mat& objectMask)
+{
+    rectangle(frame, Point(box.x, box.y), Point(box.x+box.width, box.y+box.height), Scalar(255, 178, 50), 3);
```
Inconsistent spaces, please add them: `Point(box.x + box.width, box.y + box.height)`

Daniil-Osokin commented on samples/dnn/mask_rcnn.cpp:
```diff
@@ -0,0 +1,220 @@
+    int baseLine;
+    Size labelSize = getTextSize(label, FONT_HERSHEY_SIMPLEX, 0.5, 1, &baseLine);
+    box.y = max(box.y, labelSize.height);
+    rectangle(frame, Point(box.x, box.y - (int)round(1.5*labelSize.height)), Point(box.x + (int)round(1.5*labelSize.width), box.y + baseLine), Scalar(255, 255, 255), FILLED);
```
Please, use c++ casts `static_cast<int>`.

cabelo replied:
done

Daniil-Osokin commented on samples/dnn/mask_rcnn.cpp:
```diff
@@ -0,0 +1,220 @@
+            bottom = max(0, min(bottom, frame.rows - 1));
+            Rect box = Rect(left, top, right - left + 1, bottom - top + 1);
+
+            Mat objectMask(outMasks.size[2], outMasks.size[3],CV_32F, outMasks.ptr<float>(i,classId));
```
`CV_32F` is depth, constructor needs type `CV_32FC1`.

Daniil-Osokin commented on samples/dnn/mask_rcnn.cpp:
```diff
@@ -0,0 +1,220 @@
+
+    resize(objectMask, objectMask, Size(box.width, box.height));
+    Mat mask = (objectMask > maskThreshold);
+    Mat coloredRoi = (0.3 * color + 0.7 * frame(box));
```
Suggest to get rid of extra buffers to simplify th code:
```
cv::Mat roi = frame(box);
roi = roi * 0.7 + color * 0.3;
drawContours(roi, ...);
```

cabelo replied:
done

Daniil-Osokin commented on samples/dnn/mask_rcnn.cpp:
```diff
@@ -0,0 +1,220 @@
+// Downloads:
+// http://download.tensorflow.org/models/object_detection/mask_rcnn_inception_v2_coco_2018_01_28.tar.gz
+// https://raw.githubusercontent.com/opencv/opencv_extra/master/testdata/dnn/mask_rcnn_inception_v2_coco_2018_01_28.pbtxt
+// More Information : Alessandro de Oliveira Faria (A.K.A. CABELO)- cabelo@opensuse.org
```
Imho, this line is extra here, but up to you ).

cabelo replied:
Done, sorry and thank...

cabelo commented on samples/dnn/mask_rcnn.cpp:
```diff
@@ -0,0 +1,220 @@
+#include <fstream>
+#include <sstream>
+#include <iostream>
+#include <string.h>
```
done

cabelo commented on samples/dnn/mask_rcnn.cpp:
```diff
@@ -0,0 +1,220 @@
+
+#include "common.hpp"
+
+std::string keys =
```
done

cabelo commented on samples/dnn/mask_rcnn.cpp:
```diff
@@ -0,0 +1,220 @@
+using namespace dnn;
+using namespace std;
+
+float confThreshold;
```
done

cabelo commented on samples/dnn/mask_rcnn.cpp:
```diff
@@ -0,0 +1,220 @@
+    "{ config c    |./mask_rcnn_inception_v2_coco_2018_01_28.pbtxt | iThe text graph file that has been tuned by the OpenCV’s DNN support group  }"
+    "{ cthr        | .5 | Confidence threshold. }"
+    "{ mthr        | .4 | Mask threshold. }";
+
```
done

Daniil-Osokin commented on samples/dnn/mask_rcnn.cpp:
```diff
@@ -0,0 +1,216 @@
+    Mat mask = (objectMask > maskThreshold);
+    Mat roi = frame(box);
+    roi = roi * 0.7 + color * 0.3;
+    roi.convertTo(roi, CV_8UC3);
```
Now looks better! This line, line 212 and 215 can be removed. Please, also check other comments (click on it):
![image](https://user-images.githubusercontent.com/2491976/52587103-4e710c80-2e4a-11e9-8570-d8875164c617.png)

cabelo replied:
Sorry, I did this ... I was very tired and sleepy.

Done

cabelo replied:
done all ... Sorry and thank you

sturkmen72 commented on samples/dnn/mask_rcnn.cpp:
```diff
@@ -0,0 +1,205 @@
+using namespace std;
+
+string keys =
+    "{ help  h     | | Print help message. \nUsage \n\t\t./mask_rcnn --image=logo.jpg \n\t\t ./mask_rcnn --media=teste.mp4}"
```
--media=teste.mp4

--v=teste.mp4

dkurt commented on samples/dnn/mask_rcnn.cpp:
```diff
@@ -0,0 +1,198 @@
+        colors.push_back( Scalar(rand() % 255,rand() % 255,rand() % 255 ,255.0));
+    }
+
+    Net net = readNet(modelWeights, textGraph,"tensorflow");
```
We keep the last argument only for extremal cases when `readNet` cannot determine network origin. In this example it's not necessary due it's convenient to name TensorFlow models with `.pb` extension.

cabelo replied:
ok, done

dkurt commented on samples/dnn/mask_rcnn.cpp:
```diff
@@ -0,0 +1,198 @@
+            break;
+        }
+
+        blobFromImage(frame, blob, 1.0, Size(frame.cols, frame.rows), Scalar(), true, false);
```
There are should be a command line argument to specify input blob dimensions. Take a look onto https://github.com/opencv/opencv/blob/master/samples/dnn/mask_rcnn.py.

cabelo replied:
ok, done

dkurt commented on samples/dnn/mask_rcnn.cpp:
```diff
@@ -0,0 +1,198 @@
+        putText(frame, label, Point(0, 15), FONT_HERSHEY_SIMPLEX, 0.5, Scalar(0, 0, 0));
+
+        Mat detectedFrame;
+        frame.convertTo(detectedFrame, CV_8U);
```
`frame` already has a type `CV_8U`.

tompollok commented:
@cabelo do you have your input image as shown on the top somewhere? Im having trouble to reproduce that result with your example.

![grafik](https://user-images.githubusercontent.com/4457821/53017816-20468a80-3451-11e9-9742-b05925209e33.png)
Here an output. The segmentations do not look very accurate. Id like to reproduce it with your example image if possible.

cabelo commented:
> @cabelo do you have your input image as shown on the top somewhere? Im having trouble to reproduce that result with your example.
> 
> ![grafik](https://user-images.githubusercontent.com/4457821/53017816-20468a80-3451-11e9-9742-b05925209e33.png)
> Here an output. The segmentations do not look very accurate. Id like to reproduce it with your example image if possible.

cabelo commented:
> @cabelo do you have your input image as shown on the top somewhere? Im having trouble to reproduce that result with your example.
> 
> ![grafik](https://user-images.githubusercontent.com/4457821/53017816-20468a80-3451-11e9-9742-b05925209e33.png)
> Here an output. The segmentations do not look very accurate. Id like to reproduce it with your example image if possible.

This image?

[
![img01vis](https://user-images.githubusercontent.com/675645/53020463-ace34f80-3435-11e9-8bef-a5bab8ee6f1f.jpeg)
](url)

tompollok commented:
Yes this one: It looks like this on my machine:

![grafik](https://user-images.githubusercontent.com/4457821/53021085-afa36c00-3458-11e9-9acf-a8470423ff80.png)

Are you using a different model than the one descibed in the example?

cabelo commented:
Just to know, why did not this example merge?

alalek commented on samples/dnn/mask_rcnn.cpp:
```diff
@@ -0,0 +1,203 @@
+    "{ image m     |<none>| Path to input image file.  }"
+    "{ video v     |<none>| Path to input video file.  }"
+    "{ classes n   |./mscoco_labels.names | Path to a text file with names of classes.  }"
+    "{ model w     |./mask_rcnn_inception_v2_coco_2018_01_28/frozen_inference_graph.pb | The pre-trained weights.  }"
```
`./` - do we really need this prefix?

cabelo replied:
No, sorry... fix it!

cabelo replied:
done

alalek commented on samples/dnn/mask_rcnn.cpp:
```diff
@@ -0,0 +1,203 @@
+    "{ classes n   |./mscoco_labels.names | Path to a text file with names of classes.  }"
+    "{ model w     |./mask_rcnn_inception_v2_coco_2018_01_28/frozen_inference_graph.pb | The pre-trained weights.  }"
+    "{ height H     |800|Preprocess input image by resizing to a specific height }"
+    "{ width W     |800|Preprocess input image by resizing to a specific width }"
```
fix "columns"

cabelo replied:
done

alalek commented on samples/dnn/mask_rcnn.cpp:
```diff
@@ -0,0 +1,203 @@
+    "{ model w     |./mask_rcnn_inception_v2_coco_2018_01_28/frozen_inference_graph.pb | The pre-trained weights.  }"
+    "{ height H     |800|Preprocess input image by resizing to a specific height }"
+    "{ width W     |800|Preprocess input image by resizing to a specific width }"
+    "{ config c    |./mask_rcnn_inception_v2_coco_2018_01_28.pbtxt | iThe text graph file that has been tuned by the OpenCV’s DNN support group  }"
```
- `./`
- `iThe`

cabelo replied:
done

alalek commented on samples/dnn/mask_rcnn.cpp:
```diff
@@ -0,0 +1,203 @@
+
+void drawBox(Mat& frame, int classId, float conf, Rect box, Mat& objectMask, float maskThreshold, vector<Scalar> colors, vector<string> classes);
+
+void postprocess(Mat& frame, const vector<Mat>& outs, float confThreshold, float maskThreshold, vector<Scalar> colors, vector<string> classes);
```
Helper local functions should be "static".

cabelo replied:
done

alalek commented on samples/dnn/mask_rcnn.cpp:
```diff
@@ -0,0 +1,203 @@
+    "{ cthr        | .5 | Confidence threshold. }"
+    "{ mthr        | .4 | Mask threshold. }";
+
+void drawBox(Mat& frame, int classId, float conf, Rect box, Mat& objectMask, float maskThreshold, vector<Scalar> colors, vector<string> classes);
```
`vector<>` parameters should be passed by reference (or "const reference").

cabelo replied:
done

cabelo commented:
> Yes this one: It looks like this on my machine:
> 
> ![grafik](https://user-images.githubusercontent.com/4457821/53021085-afa36c00-3458-11e9-9acf-a8470423ff80.png)
> 
> Are you using a different model than the one descibed in the example?

Are you using a different model than the one descibed in the example?
Answer: No, i used the same model.

![mask02](https://user-images.githubusercontent.com/675645/53025020-f3897780-343e-11e9-9dba-3a2c0b0704f3.png)

alalek commented on samples/dnn/mask_rcnn.cpp:
```diff
@@ -0,0 +1,203 @@
+            cout << "Image file input : " << str << endl;
+            ifstream ifile(str);
+            if (!ifile) throw("error");
+            cap.open(str);
```
Opening image via VideoCapture looks strange.
Take a look on other OpenCV samples how this case is handled.

cabelo replied:
Because i use the same object in image e video.

alalek commented on samples/dnn/mask_rcnn.cpp:
```diff
@@ -0,0 +1,203 @@
+        {
+            str = parser.get<String>("video");
+            ifstream ifile(str);
+            if (!ifile) throw("error");
```
> ifstream

"video" can be GStreamer pipeline or IP-camera URI. Don't block them.

cabelo replied:
Thank you.

alalek commented on samples/dnn/mask_rcnn.cpp:
```diff
@@ -0,0 +1,203 @@
+            if (!ifile) throw("error");
+            cap.open(str);
+        }
+        else cap.open(parser.get<int>("device"));
```
Keep arguments consistent between samples.

```.diff
-device
+camera cam
```

cabelo replied:
Done

cabelo commented:
> Just to know, why did not this example merge?

Video example bellow:
https://www.youtube.com/watch?v=0tU8991QgE8

cabelo commented:
> Original of this code is here: https://github.com/spmallick/learnopencv/blob/1f57eec12fa01017243ae57b8e1cc7eb5cbe0671/Mask-RCNN/mask_rcnn.cpp
>  Appropriate license is required for this sample (BSD compatible at least) before merge.

https://github.com/spmallick/learnopencv/blob/1f57eec12fa01017243ae57b8e1cc7eb5cbe0671/Mask-RCNN/mask_rcnn.cpp#L1

cabelo commented:
> > Original of this code is here: https://github.com/spmallick/learnopencv/blob/1f57eec12fa01017243ae57b8e1cc7eb5cbe0671/Mask-RCNN/mask_rcnn.cpp
> > Appropriate license is required for this sample (BSD compatible at least) before merge.
>

cabelo commented:
> Original of this code is here: https://github.com/spmallick/learnopencv/blob/1f57eec12fa01017243ae57b8e1cc7eb5cbe0671/Mask-RCNN/mask_rcnn.cpp
> Appropriate license is required for this sample (BSD compatible at least) before merge.

Hi,
This code is based on the first opencv example object_detection.cpp

Also uses the opencv library license. evidence below:
https://github.com/spmallick/learnopencv/blob/1f57eec12fa01017243ae57b8e1cc7eb5cbe0671/Mask-RCNN/mask_rcnn.cpp#L1

cabelo commented:
> Original of this code is here: https://github.com/spmallick/learnopencv/blob/1f57eec12fa01017243ae57b8e1cc7eb5cbe0671/Mask-RCNN/mask_rcnn.cpp
> 
> Appropriate license is required for this sample (BSD compatible at least) before merge.


Original of this code is here: https://github.com/spmallick/learnopencv/blob/1f57eec12fa01017243ae57b8e1cc7eb5cbe0671/Mask-RCNN/mask_rcnn.cpp

Appropriate license is required for this sample (BSD compatible at least) before merge.

cabelo commented:
> Original of this code is here: https://github.com/spmallick/learnopencv/blob/1f57eec12fa01017243ae57b8e1cc7eb5cbe0671/Mask-RCNN/mask_rcnn.cpp
> 
> Appropriate license is required for this sample (BSD compatible at least) before merge.

Everything is OK?

tompollok commented:
Could you tell me which open vino version you are using? Im still trying to investigate why our results look so different. Yours look fine but mine is not satisfying.

cabelo commented:
> .

I am using OpenVINO 2018 R5.01 (proprietary version)

cabelo commented:
@alalek , @dkurt all ok now?

dkurt commented on samples/dnn/mask_rcnn.cpp:
```diff
@@ -0,0 +1,204 @@
+    "{ help  h     | | Print help message. \nUsage \n\t\t./mask_rcnn --image=logo.jpg \n\t\t ./mask_rcnn --video=teste.mp4}"
+    "{ image m     |<none>| Path to input image file.  }"
+    "{ video v     |<none>| Path to input video file.  }"
+    "{ cam  C      |0| Path to input video file.  }"
```
As for me, we can keep a single argument. @cabelo, What do you think?
In example, just 

```cpp
    "{ input i     | 0 | Path to input image file or video or camera id.  }"
```

And 
```cpp
string input = parser.get<string>("input");
// ...
if (input.find_first_not_of("0123456789") == std::string::npos)  // is a camera id
  cap.open(atoi(input));
else
  cap.open(input);
```

cabelo replied:
Good idea!

cabelo replied:
done

dkurt commented on samples/dnn/mask_rcnn.cpp:
```diff
@@ -0,0 +1,204 @@
+    "{ image m     |<none>| Path to input image file.  }"
+    "{ video v     |<none>| Path to input video file.  }"
+    "{ cam  C      |0| Path to input video file.  }"
+    "{ classes n   |mscoco_labels.names | Path to a text file with names of classes.  }"
```
I think it's better to use default name `object_detection_classes_coco.txt`. This is a name of file which might be found in OpneCV's source code tree: https://github.com/opencv/opencv/blob/master/samples/data/dnn/object_detection_classes_coco.txt.

cabelo replied:
done... thank you

alalek commented:
@spmallick @sunitanyk

I'm maintainer of OpenCV source code repository.

Looks like original code for sample from this PR is from [here](https://github.com/spmallick/learnopencv/blob/1f57eec12fa01017243ae57b8e1cc7eb5cbe0671/Mask-RCNN/mask_rcnn.cpp). Unfortunately I can't find license for code in your repository.

Do you approve using (licensing) of original code and to put sample into OpenCV samples? (BSD-compatible or Apache 2)

cabelo commented:
> @spmallick @sunitanyk
> 
> I'm maintainer of OpenCV source code repository.
> 
> Looks like original code for sample from this PR is from [here](https://github.com/spmallick/learnopencv/blob/1f57eec12fa01017243ae57b8e1cc7eb5cbe0671/Mask-RCNN/mask_rcnn.cpp). Unfortunately I can't find license for code in your repository.
> 
> Do you approve using (licensing) of original code and to put sample into OpenCV samples? (BSD-compatible or Apache 2)

Hi @alalek 

The author is Satya Mallick. The repository of  @sunitanyk is Forked from Satya Mallick.

sunitanyk commented:
Hi Alexander,

Thanks for considering to add the work to OpenCV samples.

I am the author of the original code for Mask RCNN cpp using OpenCV. I have
added the license info to the file and the directory. It needs the user to
add attribution to Big Vision LLC (LearnOpenCV.com) in the new sample file.
The license is BSD-3-Clause-Attribution.

The updated files are at:

https://github.com/spmallick/learnopencv/blob/master/Mask-RCNN/mask_rcnn.cpp

https://github.com/spmallick/learnopencv/blob/master/Mask-RCNN/LICENSE

Hope this helps !

Thanks,
Sunita







On Sat, Mar 9, 2019 at 8:20 AM Alessandro de Oliveira Faria (A.K.A.CABELO) <
notifications@github.com> wrote:

> @spmallick <https://github.com/spmallick> @sunitanyk
> <https://github.com/sunitanyk>
>
> I'm maintainer of OpenCV source code repository.
>
> Looks like original code for sample from this PR is from here
> <https://github.com/spmallick/learnopencv/blob/1f57eec12fa01017243ae57b8e1cc7eb5cbe0671/Mask-RCNN/mask_rcnn.cpp>.
> Unfortunately I can't find license for code in your repository.
>
> Do you approve using (licensing) of original code and to put sample into
> OpenCV samples? (BSD-compatible or Apache 2)
>
> Hi @alalek <https://github.com/alalek>
>
> The author is Satya Mallick. The repository of @sunitanyk
> <https://github.com/sunitanyk> is Forked from Satya Mallick.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/opencv/opencv/pull/13771#issuecomment-471197034>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AaAf0ZXswcVEz7wRWb_RD6OmbIJHcsDxks5vU99AgaJpZM4amgDM>
> .
>

cabelo commented:
> Hi Alexander, Thanks for considering to add the work to OpenCV samples. I am the author of the original code for Mask RCNN cpp using OpenCV. I have added the license info to the file and the directory. It needs the user to add attribution to Big Vision LLC (LearnOpenCV.com) in the new sample file. The license is BSD-3-Clause-Attribution. The updated files are at: https://github.com/spmallick/learnopencv/blob/master/Mask-RCNN/mask_rcnn.cpp https://github.com/spmallick/learnopencv/blob/master/Mask-RCNN/LICENSE Hope this helps ! Thanks, Sunita
> […](#)
> On Sat, Mar 9, 2019 at 8:20 AM Alessandro de Oliveira Faria (A.K.A.CABELO) < ***@***.***> wrote: @spmallick <https://github.com/spmallick> @sunitanyk <https://github.com/sunitanyk> I'm maintainer of OpenCV source code repository. Looks like original code for sample from this PR is from here <https://github.com/spmallick/learnopencv/blob/1f57eec12fa01017243ae57b8e1cc7eb5cbe0671/Mask-RCNN/mask_rcnn.cpp>. Unfortunately I can't find license for code in your repository. Do you approve using (licensing) of original code and to put sample into OpenCV samples? (BSD-compatible or Apache 2) Hi @alalek <https://github.com/alalek> The author is Satya Mallick. The repository of @sunitanyk <https://github.com/sunitanyk> is Forked from Satya Mallick. — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub <[#13771 (comment)](https://github.com/opencv/opencv/pull/13771#issuecomment-471197034)>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AaAf0ZXswcVEz7wRWb_RD6OmbIJHcsDxks5vU99AgaJpZM4amgDM> .



> Hi Alexander, Thanks for considering to add the work to OpenCV samples. I am the author of the original code for Mask RCNN cpp using OpenCV. I have added the license info to the file and the directory. It needs the user to add attribution to Big Vision LLC (LearnOpenCV.com) in the new sample file. The license is BSD-3-Clause-Attribution. The updated files are at: https://github.com/spmallick/learnopencv/blob/master/Mask-RCNN/mask_rcnn.cpp https://github.com/spmallick/learnopencv/blob/master/Mask-RCNN/LICENSE Hope this helps ! Thanks, Sunita
> […](#)
> On Sat, Mar 9, 2019 at 8:20 AM Alessandro de Oliveira Faria (A.K.A.CABELO) < ***@***.***> wrote: @spmallick <https://github.com/spmallick> @sunitanyk <https://github.com/sunitanyk> I'm maintainer of OpenCV source code repository. Looks like original code for sample from this PR is from here <https://github.com/spmallick/learnopencv/blob/1f57eec12fa01017243ae57b8e1cc7eb5cbe0671/Mask-RCNN/mask_rcnn.cpp>. Unfortunately I can't find license for code in your repository. Do you approve using (licensing) of original code and to put sample into OpenCV samples? (BSD-compatible or Apache 2) Hi @alalek <https://github.com/alalek> The author is Satya Mallick. The repository of @sunitanyk <https://github.com/sunitanyk> is Forked from Satya Mallick. — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub <[#13771 (comment)](https://github.com/opencv/opencv/pull/13771#issuecomment-471197034)>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AaAf0ZXswcVEz7wRWb_RD6OmbIJHcsDxks5vU99AgaJpZM4amgDM> .

@sunitanyk , thank you very much for the attention and congratulations for the work.

cabelo commented:
@sunitanyk, thank you for the attention and congratulations for the work.

Thank you again for adding the license:  BSD-3-Clause-Attribution

sturkmen72 commented on samples/dnn/mask_rcnn.cpp:
```diff
@@ -0,0 +1,200 @@
+// Deep learning based Object Detection and Instance Segmentation using Mask R-CNN
+// VIDEO 360 DEMO: https://www.youtube.com/watch?v=0tU8991QgE8
```
(my personal opinion) i do not want to see this kind of links in the code. i think video links should be only in documentation and devs should be selective and refuse low quality videos.

sunitanyk replied:
I had similar opinion about the linked video too. It would be great if you could link in the code to the blog https://www.learnopencv.com/deep-learning-based-object-detection-and-instance-segmentation-using-mask-r-cnn-in-opencv-python-c/   
It explains its usage and has a good results video which was created after hours of effort on different kinds of objects.

cabelo replied:
@sunitanyk  and @sturkmen72 I removed the video from the code.
Thanks for the suggestion.

You can use the video https://www.youtube.com/watch?v=0tU8991QgE8 on the blog. It is a pleasure to contribute to your excellent work.

sunitanyk commented:
I had a similar opinion about the linked video too. It would be great if
you could link in the code to the blog
https://www.learnopencv.com/deep-learning-based-object-detection-and-instance-segmentation-using-mask-r-cnn-in-opencv-python-c/
It explains its usage and has a good results video which was created after
hours of effort on different kinds of objects.

On Sat, Mar 9, 2019 at 4:08 PM Suleyman TURKMEN <notifications@github.com>
wrote:

> *@sturkmen72* commented on this pull request.
> ------------------------------
>
> In samples/dnn/mask_rcnn.cpp
> <https://github.com/opencv/opencv/pull/13771#discussion_r264019007>:
>
> > @@ -0,0 +1,200 @@
> +// Deep learning based Object Detection and Instance Segmentation using Mask R-CNN
> +// VIDEO 360 DEMO: https://www.youtube.com/watch?v=0tU8991QgE8
>
> (my personal opinion) i do not want to see this kind of links in the code.
> i think video links should be only in documentation and devs should be
> selective and refuse low quality videos.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/opencv/opencv/pull/13771#pullrequestreview-212582213>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AaAf0QiZ7HbS4Iphp9lCd8fh7ETPgy3wks5vVE0LgaJpZM4amgDM>
> .
>

cabelo commented:
> I had a similar opinion about the linked video too. It would be great if you could link in the code to the blog https://www.learnopencv.com/deep-learning-based-object-detection-and-instance-segmentation-using-mask-r-cnn-in-opencv-python-c/ It explains its usage and has a good results video which was created after hours of effort on different kinds of objects.
> […](#)
> On Sat, Mar 9, 2019 at 4:08 PM Suleyman TURKMEN ***@***.***> wrote: ***@***.**** commented on this pull request. ------------------------------ In samples/dnn/mask_rcnn.cpp <[#13771 (comment)](https://github.com/opencv/opencv/pull/13771#discussion_r264019007)>: > @@ -0,0 +1,200 @@ +// Deep learning based Object Detection and Instance Segmentation using Mask R-CNN +// VIDEO 360 DEMO: https://www.youtube.com/watch?v=0tU8991QgE8 (my personal opinion) i do not want to see this kind of links in the code. i think video links should be only in documentation and devs should be selective and refuse low quality videos. — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub <[#13771 (review)](https://github.com/opencv/opencv/pull/13771#pullrequestreview-212582213)>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AaAf0QiZ7HbS4Iphp9lCd8fh7ETPgy3wks5vVE0LgaJpZM4amgDM> .

Removed, thanks for the suggestion.

cabelo commented:
Removed, thanks for the suggestion.

sturkmen72 commented on samples/dnn/mask_rcnn.cpp:
```diff
@@ -0,0 +1,199 @@
+
+string keys =
+    "{ help  h     | | Print help message. \nUsage \n\t\t./mask_rcnn --image=logo.jpg \n\t\t ./mask_rcnn --video=teste.mp4}"
+    "{ image m     |<none>| Path to input image file.  }"
```
please remove this line

sturkmen72 commented on samples/dnn/mask_rcnn.cpp:
```diff
@@ -0,0 +1,199 @@
+    "{ image m     |<none>| Path to input image file.  }"
+    "{ input i     | 0 | Path to input image file or video or camera id.  }"
+    "{ classes n   |object_detection_classes_coco.txt| Path to a text file with names of classes.  }"
+    "{ model w     |mask_rcnn_inception_v2_coco_2018_01_28/frozen_inference_graph.pb | The pre-trained weights.  }"
```
`model m` is more suitable ( also w is for width)

cabelo replied:
done

sturkmen72 commented on samples/dnn/mask_rcnn.cpp:
```diff
@@ -0,0 +1,199 @@
+    "{ input i     | 0 | Path to input image file or video or camera id.  }"
+    "{ classes n   |object_detection_classes_coco.txt| Path to a text file with names of classes.  }"
+    "{ model w     |mask_rcnn_inception_v2_coco_2018_01_28/frozen_inference_graph.pb | The pre-trained weights.  }"
+    "{ height H    |800| Preprocess input image by resizing to a specific height }"
```
height h

cabelo replied:
done

sturkmen72 commented on samples/dnn/mask_rcnn.cpp:
```diff
@@ -0,0 +1,199 @@
+    "{ classes n   |object_detection_classes_coco.txt| Path to a text file with names of classes.  }"
+    "{ model w     |mask_rcnn_inception_v2_coco_2018_01_28/frozen_inference_graph.pb | The pre-trained weights.  }"
+    "{ height H    |800| Preprocess input image by resizing to a specific height }"
+    "{ width W     |800| Preprocess input image by resizing to a specific columns }"
```
width w

cabelo replied:
done

sturkmen72 commented on samples/dnn/mask_rcnn.cpp:
```diff
@@ -0,0 +1,199 @@
+using namespace std;
+
+string keys =
+    "{ help  h     | | Print help message. \nUsage \n\t\t./mask_rcnn --image=logo.jpg \n\t\t ./mask_rcnn --video=teste.mp4}"
```
`"{ help  h     | | Print help message. }"` enough

sturkmen72 commented on samples/dnn/mask_rcnn.cpp:
```diff
@@ -0,0 +1,199 @@
+{
+    CommandLineParser parser(argc, argv, keys);
+    parser.about("This sample demonstrates instance segmentation network called Mask-RCNN.");
+    if (argc == 1 || parser.has("help"))
```
`if (parser.has("help"))` enough here because default values for params are all defined

sturkmen72 commented on samples/dnn/mask_rcnn.cpp:
```diff
@@ -0,0 +1,199 @@
+    "{ help  h     | | Print help message. \nUsage \n\t\t./mask_rcnn --image=logo.jpg \n\t\t ./mask_rcnn --video=teste.mp4}"
+    "{ image m     |<none>| Path to input image file.  }"
+    "{ input i     | 0 | Path to input image file or video or camera id.  }"
+    "{ classes n   |object_detection_classes_coco.txt| Path to a text file with names of classes.  }"
```
"{ classes     |mscoco_labels.names| Path to a text file with names of classes.  }"

cabelo replied:
done

sturkmen72 commented on samples/dnn/mask_rcnn.cpp:
```diff
@@ -0,0 +1,199 @@
+    "{ model  m    |mask_rcnn_inception_v2_coco_2018_01_28/frozen_inference_graph.pb | The pre-trained weights.  }"
+    "{ height h    |800| Preprocess input image by resizing to a specific height }"
+    "{ width w     |800| Preprocess input image by resizing to a specific columns }"
+    "{ config c    |mask_rcnn_inception_v2_coco_2018_01_28.pbtxt |The text graph file that has been tuned by the OpenCV’s DNN support group  }"
```
`OpenCV’s`->OpenCV's

still `height h` and `help  h`

cabelo replied:
Sorry, done...

sturkmen72 commented on samples/dnn/mask_rcnn.cpp:
```diff
@@ -0,0 +1,199 @@
+    roi = roi * 0.7 + color * 0.3;
+    roi.convertTo(roi, CV_8UC3);
+
+    vector<Mat> contours;
```
consider the following change
```
    vector<vector<Point>> contours;
    findContours(mask, contours, RETR_TREE, CHAIN_APPROX_SIMPLE);
    polylines(roi, contours, true, color, 2, LINE_8);
```

cabelo replied:
Excellent ... done!

cabelo replied:
done

alalek commented:
@sunitanyk Thank you an update! Sorry for bothering you.

Closing this for now, because there are no plans to restrict OpenCV licensing.

---

@cabelo Next time, please ensure that you wrote code by yourself, or you using code with appropriate license.

cabelo commented:
> @sunitanyk Thank you an update! Sorry for bothering you.
> 
> Closing this for now, because there are no plans to restrict OpenCV licensing.
> 
> @cabelo Next time, please ensure that you wrote code by yourself, or you using code with appropriate license.

Sorry for the inconvenience, I'll rewrite Mask R-CNN from the beginning.

thank you
